### PR TITLE
vmalert: improve performances when rules produce large volumes of results

### PR DESCRIPTION
**Pull request by Haleygo**
1. Avoid storing the last evaluation results outside of rules, check for stale time series as soon as possible;
2. remove duplicated template `Clone()`.

This pull request is primarily reducing memory usage when rules produce large volumes of results, as seen in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6894. 
The CPU time spent on garbage collection remains high and may be addressed in a separate PR.

Below are resource usages comparison between v1.105.0:
image

### CODE REVIEW

**Review by hagen1778 — COMMENTED** [review]
Good idea about stale series, and significantly reduces amount of mem and allocations!
A few comments though

File: app/vmalert/rule/alerting.go [hunk 1]
```diff
@@ -454,13 +456,17 @@ func (ar *AlertingRule) exec(ctx context.Context, ts time.Time, limit int) ([]pr
 		ar.logDebugf(ts, a, "created in state PENDING")
 	}
 	var numActivePending int
+	// store alerts' labels which are `FIRING => INACTIVE`, `PENDING => INACTIVE` or `PENDING => FIRING` in this iteration,
+	// need to create stale time series for them later.
+	var pendingToFiring, pendingToInactive, firingToInactive []map[string]string
```
[the remark this answers is not in the document]
Comment (Haleygo) [reply app/vmalert/rule/alerting.go]: Cleared the code in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7390/commits/9e79fb2e1156617b56284323e19554486e51898f, please check.

File: app/vmalert/rule/group_test.go [hunk 2]
```diff
@@ -383,153 +380,6 @@ func TestGetResolveDuration(t *testing.T) {
 	f(2*time.Minute, 0, 1*time.Minute, 8*time.Minute)
 }
 
-func TestGetStaleSeries(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Haleygo) [reply app/vmalert/rule/group_test.go:386]: The generated stale timeseries are tested during rule exec, see https://github.com/VictoriaMetrics/VictoriaMetrics/blob/9e79fb2e1156617b56284323e19554486e51898f/app/vmalert/rule/alerting_test.go#L207 and https://github.com/VictoriaMetrics/VictoriaMetrics/blob/9e79fb2e1156617b56284323e19554486e51898f/app/vmalert/rule/recording_test.go#L113.

**Review by hagen1778 — APPROVED** [review]
LGTM

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!

**Comment by hagen1778** [comment]
@Haleygo this change is missing note in the changelog. Would you mind adding one?

**Comment by Haleygo** [comment]
> @Haleygo this change is missing note in the changelog. Would you mind adding one?

@hagen1778 I will, didn't add it cause it's not a use-facing issue.

**Pull request merged**